core/src/main/java/org/apache/iceberg/io/ContentCache.java

```diff
@@ -0,0 +1,286 @@
+  }
+
+  public void invalidateAll() {
+    cache.invalidateAll();
+  }
+
+  public void cleanUp() {
+    cache.cleanUp();
+  }
+
+  public long estimatedCacheSize() {
+    return cache.estimatedSize();
+  }
+
+  @Override
+  public String toString() {
+    return getClass().getSimpleName()
+        + '{'
+        + "expireAfterAccessMs="
+        + expireAfterAccessMs
+        + ", "
+        + "maxContentLength="
+        + maxContentLength
+        + ", "
+        + "maxTotalBytes="
+        + maxTotalBytes
+        + ", "
+        + "cacheStats="
+        + cache.stats()
+        + '}';
+  }
+
+  private static class CacheEntry {
+    private final long length;
+    private final List<ByteBuffer> buffers;
+
+    private CacheEntry(long length, List<ByteBuffer> buffers) {
+      this.length = length;
+      this.buffers = buffers;
+    }
+  }
+
+  private static class CachingInputFile implements InputFile {
+    private final ContentCache contentCache;
+    private final FileIO io;
+    private final String location;
+    private final long length;
+    private InputFile fallbackInputFile = null;
+
+    private CachingInputFile(ContentCache cache, FileIO io, String location, long length) {
+      this.contentCache = cache;
+      this.io = io;
+      this.location = location;
+      this.length = length;
+    }
+
+    private InputFile wrappedInputFile() {
+      if (fallbackInputFile == null) {
+        fallbackInputFile = io.newInputFile(location, length);
+      }
+      return fallbackInputFile;
+    }
+
+    @Override
+    public long getLength() {
+      CacheEntry buf = contentCache.getIfPresent(location);
+      return (buf != null) ? buf.length : wrappedInputFile().getLength();
+    }
+
+    /**
+     * Opens a new {@link SeekableInputStream} for the underlying data file, either through cache or
+     * through the inner FileIO.
+     *
+     * <p>If data file is not cached yet, and it can fit in the cache, the file content will be
+     * cached first before returning a {@link ByteBufferInputStream}. Otherwise, return a new
+     * SeekableInputStream from the inner FIleIO.
+     *
+     * @return a {@link ByteBufferInputStream} if file exist in the cache or can fit in the cache.
+     *     Otherwise, return a new SeekableInputStream from the inner FIleIO.
+     */
+    @Override
+    public SeekableInputStream newStream() {
+      try {
+        // read-through cache if file length is less than or equal to maximum length allowed to
+        // cache.
+        if (getLength() <= contentCache.maxContentLength()) {
+          return cachedStream();
+        }
+
+        // fallback to non-caching input stream.
+        return wrappedInputFile().newStream();
+      } catch (FileNotFoundException e) {
+        throw new NotFoundException(e, "Failed to open input stream for file: %s", location);
+      } catch (IOException e) {
+        throw new RuntimeIOException(e, "Failed to open input stream for file: %s", location);
```

Comment:
could you add e.toString to the error ext, as there's always the risk that something (hive...) will not print the innermost exception after wrapping things a few more times.

Reply:
Done in [f76fa5d](https://github.com/apache/iceberg/pull/4518/commits/f76fa5d7f8257c63fd90df32fd841ac989cb26af).